Supporting invocations/await for the OneOf-Value

dmuehlenhoff opened the pull request:
Hi,

first of all thanks for making this suppressor. It really makes OneOfs more usable.

I added support for method calls and await expressions to be suppressable. When using this suppressor I often run into the case that we do the following

```csharp
var result = await SomethingAsync();
return result.Value switch
{
    ...
};
```

Less often, we just need to call a method without await. This PR adds support for both cases. I added simple test cases for both (invocation and await) to both suppressors.

```csharp
return (await SomethingAsync()).Value switch
{
    ...
};
```

This also makes nested switch expressions work without (local) functions, which weren't possible before since you never could declare result. I also added a test for this (more as a documentation thing than an actual "does this really work" test)

```csharp
return SomeFunc().Value switch
{
    SomeError => ...,
    Success s => SomeOtherFunc(s).Value switch
    {
       ...
    }
};
```

I moved some common logic for determining the ITypeSymbol of the expression to another helper because it's a bit more code now and inlining that into both suppressors seemed not to be the greatest idea.

It would be great if you could incorporate these changes into a new version of the published package. If you need me to change stuff just tell me what. You can also edit this branch yourself.

Thanks in advance,
David

Commit:
Logic + Switch statement

Commit:
Switch-expression

Commit:
Nested switch expression test

shuebner commented:
Hi David,
thank you for your contribution. It looks good on first glance.

I will get around to reviewing it over the weekend.
I also remember there being some issues around licensing and who owns contributions. I will check those too and may have to require you to open a new PR.

shuebner commented on OneOfDiagnosticSuppressor/ExpressionHelper.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        return GetUnparenthesizedExpression(awaitSyntax.Expression) switch
+        {
+            IdentifierNameSyntax id => GetTypeSymbolFromIdentifier(model, id),
```
It would be nice to tease out this case with a test.

shuebner commented on OneOfDiagnosticSuppressor/ExpressionHelper.cs:
```diff
@@ -0,0 +1,67 @@
+        return GetUnparenthesizedExpression(awaitSyntax.Expression) switch
+        {
+            IdentifierNameSyntax id => GetTypeSymbolFromIdentifier(model, id),
+            MemberAccessExpressionSyntax { Name: IdentifierNameSyntax id } => GetTypeSymbolFromIdentifier(model, id),
```
It would be nice to tease out this case with a test.

shuebner commented on OneOfDiagnosticSuppressor/ExpressionHelper.cs:
```diff
@@ -0,0 +1,67 @@
+            IdentifierNameSyntax id => GetTypeSymbolFromIdentifier(model, id),
+            MemberAccessExpressionSyntax { Name: IdentifierNameSyntax id } => GetTypeSymbolFromIdentifier(model, id),
+            InvocationExpressionSyntax invocation => GetTypeSymbolFromInvocation(model, invocation, expectTaskReturnType: true),
+            _ => null,
```
There could conceivably be an `AwaitExpressionSyntax` as well, think `Task.Factory.StartNew(() => DoWorkAsync())`.
This is a niche case of course, but while we are at it... ;-)

shuebner commented on OneOfDiagnosticSuppressor/ExpressionHelper.cs:
```diff
@@ -0,0 +1,67 @@
+    static ITypeSymbol? GetTypeSymbolFromInvocation(SemanticModel model, InvocationExpressionSyntax invocation, bool expectTaskReturnType)
+    {
+        var invokedSymbol = model.GetSymbolInfo(invocation);
+        if (invokedSymbol.Symbol is not IMethodSymbol methodSymbol || methodSymbol.ReturnType is not INamedTypeSymbol returnTypeSymbol)
```
Could be simplified to `invokedSymbol.Symbol is not IMethodSymbol { ReturnType: INamedTypeSymbol returnTypeSymbol }`

Commit:
Changes for PR
- Refactored await and non-await syntax to use the same logic
- Added missing tests

dmuehlenhoff commented:
I have updated the PR. I didn't only include nested awaits, but refactored the logic to use the same expression analyzer for awaits and non-awaits (so when, e.g. support for `(await arr[0]).Value` is added, support for `arr[0].Value` is also automatically added (or vice-versa)

shuebner commented:
@dmuehlenhoff Thank you for your changes.

I will do a bit of formatting and renaming (e. g. `switchSyntaxExpression` which is really the switchee).
I also found that it is not necessary to distinguish between the expression types just to get the type on which `Value` was called.
Only the await expression seems to require special handling (and only when the `Task` is in a `var` declared variable) to get the true nullable annotations of the generic type arguments on `OneOf<...>`.

shuebner committed:
refactor: simplify getting type from expression

shuebner committed:
refactor: rename and format
